mesonbuild/compilers/detect.py

```diff
@@ -1034,7 +1034,7 @@ def detect_rust_compiler(env: 'Environment', for_machine: MachineChoice) -> Rust
                     extra_args['machine'] = cc.linker.machine
                 else:
                     exelist = cc.linker.exelist + cc.linker.get_always_args()
-                    if 'ccache' in exelist[0]:
+                    if os.path.basename(exelist[0]) == 'ccache':
```

Comment:
Good catch! There's an interesting quirk about this heuristic.

This code is ancient, from 2020. In 2022, we added support to the compiler classes themselves, to track the command line with and without ccache (they are passed individually and combined into self.exelist).

So we can do `cc.exelist_no_ccache` but this doesn't work for linkers as those get inited with `cc.exelist`.

I wonder if we should track this in the linker the same way, and then use that directly instead of checking basename and pruning the first element.

Checking for an exact match is going to fail to detect sccache, by the way -- that's probably why the original implementation tried to use `in` on the whole path. The solution is trivial, combine both approaches and check for `'ccache' in os.path.basename(exelist[0])`.

Reply:
I'd simply check for `scache` too then, as otherwise anything with "ccache" in the name will be removed. I have been in situations where, for whatever reason, we had a `ccache-cc` script for example.

This also seems to be what is done when calculating `exelist_no_ccache`.

https://github.com/mesonbuild/meson/blob/cf5adf0c646474f0259d123fad60ca5ed38ec891/mesonbuild/envconfig.py#L433-L445